### DISCUSSION TITLE
post/high-performance-crawler/

### DISCUSSION THREAD

**Discussion by giscus** [lyc8503/lyc8503#32] [Announcements]
# post/high-performance-crawler/

.admonition {   margin: .75em 0;   padding: .6rem;   overflow: hidden;   font-size: 12px;   page-break-inside: avoid;   border-left: .3rem solid #42b983;   border-radius: .3rem;   box-shadow: 0 0.1re

https://blog.lyc8503.net/post/high-performance-crawler/

**Comment by oxtiger** [comment]
好奇你爬的啥站呀

**Reply by lyc8503** [reply]
可以推测出是个反爬比较宽松的大站, 不过很多站点都是一样的套路.

**Comment by Sduby22** [comment]
咦，要是复用http session的话为什么outbound ip还可以随机选择呢，为啥不是一个tcp连接就固定了一个ip地址嘛？

**Reply by lyc8503** [reply]
Session 在这里是一个连接池，内部会[维护若干 TCP 连接](https://docs.aiohttp.org/en/stable/client_reference.html#connectors)，每个连接固定一个 IP 地址，默认配置下 QPS 平摊到每个连接就不会触发 rate limit。这里我确实没有细究，不过观察到这个连接池还会每过一段时间断开部分旧连接并创建新的连接，此时 IP 也会变化。

**Reply by Sduby22** [reply]
感谢回复，我个人理解session维护连接池的逻辑和我们 if hit limit then switch ip + load balancing的需求恰好重合了。实际如果自己写连接池的话可能需要处理下如果返回类似429的rate limit错误，就说明这个连接不能用了，换新连接重试之类的逻辑。ipv6 freedom切换落地ip这个真的是学到了，感谢博主

**Comment by aicorein** [comment]
非常好思路，使我异步事件循环快速旋转
